pkg/controller/stacks/stack/stack.go

```diff
@@ -160,29 +188,169 @@ func (h *stackHandler) update(ctx context.Context) (reconcile.Result, error) {
+	// TODO(displague) refine 'editor' persona roles
+	for persona := range roleVerbs {
+		labelsCopy := map[string]string{}
+		for k, v := range labels {
+			labelsCopy[k] = v
+		}
+
+		rules := []rbacv1.PolicyRule{}
+		for _, crd := range h.ext.Spec.CRDs {
+			rules = append(rules, rbacv1.PolicyRule{
+				APIGroups: []string{crd.GroupVersionKind().Group},
+				Resources: []string{crd.Kind},
+				Verbs:     roleVerbs[persona],
+			})
+		}
+		name := fmtPersonaRole(h.ext, persona)
+
+		var crossplaneScope string
+
+		if h.isNamespaced() {
+			crossplaneScope = "namespace"
+		} else {
+			crossplaneScope = "environment"
+		}
+
+		aggregationLabel := fmt.Sprintf(labelAggregateFmt, crossplaneScope, persona)
+
+		cr := &rbacv1.ClusterRole{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   name,
+				Labels: labelsCopy,
+			},
+			Rules: rules,
+		}
+
+		meta.AddLabels(cr, map[string]string{
+			aggregationLabel: "true",
+		})
+
+		meta.AddLabels(cr, stacks.ParentLabels(h.ext))
+
+		if h.isNamespaced() {
+			meta.AddLabels(cr, map[string]string{
+				fmt.Sprintf(labelNamespaceFmt, h.ext.GetNamespace()): "true",
+			})
+		}
+
+		if err := h.kube.Create(ctx, cr); err != nil && !kerrors.IsAlreadyExists(err) {
+			return errors.Wrap(err, "failed to create persona cluster roles")
+		}
+	}
+	return nil
+}
+
+func generateNamespaceClusterRoles(stack *v1alpha1.Stack) (roles []*rbacv1.ClusterRole) {
+	personas := []string{"admin", "edit", "view"}
+
+	namespaced := (stack.Spec.PermissionScope == string(apiextensions.NamespaceScoped))
+	if !namespaced {
+		return
+	}
+
+	nsName := stack.GetNamespace()
+
+	for _, persona := range personas {
+		name := fmt.Sprintf("crossplane:ns:%s:%s", nsName, persona)
+
+		labels := map[string]string{
+			fmt.Sprintf(labelNamespaceFmt, nsName): "true",
+			labelScope:                             "namespace",
+		}
+
+		if persona == "admin" {
+			labels[fmt.Sprintf(labelAggregateFmt, "crossplane", persona)] = "true"
+		}
+
+		role := &rbacv1.ClusterRole{
+			AggregationRule: &rbacv1.AggregationRule{
+				ClusterRoleSelectors: []metav1.LabelSelector{
+					{
+						MatchLabels: map[string]string{
+							fmt.Sprintf(labelAggregateFmt, "namespace", persona): "true",
+							fmt.Sprintf(labelNamespaceFmt, nsName):               "true",
+						},
+					},
+					{
+						MatchLabels: map[string]string{
+							fmt.Sprintf(labelAggregateFmt, "namespace-default", persona): "true",
+						},
+					},
+				},
+			},
+
+			// TODO(displague) set parent labels?
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   name,
+				Labels: labels,
+			},
+		}
+
+		roles = append(roles, role)
+	}
+
+	return roles
+}
+
+func (h *stackHandler) createNamespaceClusterRoles(ctx context.Context) error {
+	ns := &corev1.Namespace{}
+	nsName := h.ext.GetNamespace()
+
+	if err := h.kube.Get(ctx, types.NamespacedName{Name: nsName}, ns); err != nil {
+		return errors.Wrapf(err, "failed to get namespace %q for stackinstall %q", nsName, h.ext.GetName())
```

Comment:
should this error message should say `stack` instead of `stackinstall`? (same for the error message below if creating a clusterrole fails)